lib/rucio/daemons/conveyor/stager.py

```diff
@@ -92,23 +86,25 @@ def stager(once=False, rses=[], mock=False,
         max_time_in_queue['default'] = 168
     logging.debug("Maximum time in queue for different activities: %s" % max_time_in_queue)
 
+    activity_next_exe_time = defaultdict(time.time)
     executable = ' '.join(sys.argv)
     hostname = socket.getfqdn()
     pid = os.getpid()
     hb_thread = threading.current_thread()
     heartbeat.sanity_check(executable=executable, hostname=hostname)
-    hb = heartbeat.live(executable, hostname, pid, hb_thread)
+    heart_beat = heartbeat.live(executable, hostname, pid, hb_thread)
+    prepend_str = 'Thread [%i/%i] : ' % (heart_beat['assign_thread'] + 1, heart_beat['nr_threads'])
+    logging.info(prepend_str + 'Stager starting with bring_online %s seconds' % (bring_online))
 
-    logging.info('Stager started - process (%i/%i) thread (%i/%i)' % (process, total_processes,
-                                                                      hb['assign_thread'], hb['nr_threads']))
-
-    threadPool = ThreadPool(total_threads)
-    activity_next_exe_time = defaultdict(time.time)
+    time.sleep(10)  # To prevent running on the same partition if all the poller restart at the same time
```

Comment:
random time instead?

Reply:
Isn’t that a classic race condition?

Reply:
In principle it is, but there are other mechanisms (row-locking) which ensure that only one worker works at a certain request at a time. This timeout here should just ensure that it doesn't even come to this conflicts on start-up (daily restart etc.). But remember, these daemons can crash/startup at any time, so it can always happen that at some point in time the work partitions overlap until the end of the loop and the next heartbeat resync (line 107) - that's why there are the row-locking mechanisms as well.

Reply:
Still, arbitrary sleeps catch a lot of attention (why 10 seconds?  why not 1?  why not 100?  etc).

How about using a barrier with a timeout?

https://docs.python.org/3/library/threading.html#barrier-objects (note that implementation is Python 3.x only but the implementation can be found in any OS book in the library...)

That will allow startup proceed as fast as possible -- but limit the lag possible from one thread.

Reply:
The 10s are just a result of experience. The barrier-object will work within the process, but what we want to do here is to "sync" up multiple workers running on different machines, which can't be done with the barrier-object.
It's really just an artefact of the "synchronised" restart every day. The code would work without the sleep, there would just be a few row-lock conflicts in the first iteration of the loop.